src/Middleware/RequestLimiter/src/RequestLimiterMiddleware.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.ResourceLimits;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.RequestLimiter
+{
+    public class RequestLimiterMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ILogger _logger;
+        private readonly RequestLimiterOptions _options;
+
+        public RequestLimiterMiddleware(RequestDelegate next, ILoggerFactory loggerFactory, IOptions<RequestLimiterOptions> options)
+        {
+            _next = next;
+            _logger = loggerFactory.CreateLogger<RequestLimiterMiddleware>();
+            _options = options.Value;
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            _logger.LogInformation("Resource limiting: " + context.Request.Path);
+
+            var endpoint = context.GetEndpoint();
+            var attributes = endpoint?.Metadata.GetOrderedMetadata<RequestLimitAttribute>();
+
+            if (attributes == null)
+            {
+                await _next.Invoke(context);
```

Comment:
Fast path avoid the async/await unless there are attributes.

Reply:
FYI I haven't been too careful with perf in terms of implementation yet.